substrate/validator-sets/pallet/src/lib.rs

```diff
@@ -0,0 +1,206 @@
+    pub participants: Vec<T::AccountId>,
+  }
+
+  #[cfg(feature = "std")]
+  impl<T: Config> Default for GenesisConfig<T> {
+    fn default() -> Self {
+      GenesisConfig { bond: Amount(0), coins: Coin(0), participants: vec![] }
+    }
+  }
+
+  // Max of 16 coins per validator set
+  // At launch, we'll have BTC, ETH, DAI, and XMR
+  // In the future, these will be split into separate validator sets, so we're already not
+  // planning expansion beyond just a few coins per validator set
+  // The only case which really makes sense for multiple coins in a validator set is:
+  // 1) The coins are small, easy to run, and make no sense to be in their own set
+  // In this case, it's still hard to ask validators to run 16 different nodes
+  // 2) The coins are all on the same network yet there's no DEX on-chain
+  // In these cases, it'd be hard to find and justify 16 different coins from that single chain
+  // This could probably be just 8, yet 16 is a hedge for the unforseen
+  // If necessary, this can be increased with a fork
+  type MaxCoinsPerSet = ConstU32<16>;
+
+  // Support keys up to 96 bytes (BLS12-381 G2)
+  const MAX_KEY_LEN: u32 = 96;
+  type MaxKeyLen = ConstU32<MAX_KEY_LEN>;
+
+  #[derive(Clone, PartialEq, Eq, Debug, Encode, Decode, TypeInfo, MaxEncodedLen)]
+  pub struct ValidatorSet<T: Config> {
+    bond: Amount,
+    coins: BoundedVec<Coin, MaxCoinsPerSet>,
+
+    // Participant and their amount bonded to this set
+    // Limit each set to 100 participants for now
+    participants: BoundedVec<(T::AccountId, Amount), ConstU32<100>>,
+  }
+
+  #[pallet::pallet]
+  #[pallet::generate_store(pub(super) trait Store)]
+  pub struct Pallet<T>(PhantomData<T>);
+
+  /// The details of a validator set instance.
+  #[pallet::storage]
+  #[pallet::getter(fn validator_set)]
+  pub type ValidatorSets<T: Config> =
+    StorageMap<_, Twox64Concat, ValidatorSetInstance, ValidatorSet<T>, OptionQuery>;
+
+  type Key = BoundedVec<u8, MaxKeyLen>;
+
+  /// The key for a given validator set instance coin.
+  #[pallet::storage]
+  #[pallet::getter(fn key)]
+  pub type Keys<T: Config> =
+    StorageMap<_, Twox64Concat, (ValidatorSetInstance, Coin), Key, OptionQuery>;
+
+  /// If an account has voted for a specific key or not. Prevents them from voting multiple times.
+  #[pallet::storage]
+  #[pallet::getter(fn voted)]
+  pub type Voted<T: Config> = StorageMap<_, Blake2_128Concat, (T::AccountId, Key), (), OptionQuery>;
+
+  /// How many times a key has been voted for. Once consensus is reached, the keys will be adopted.
+  #[pallet::storage]
+  #[pallet::getter(fn vote_count)]
+  pub type VoteCount<T: Config> =
+    StorageMap<_, Blake2_128Concat, (ValidatorSetInstance, Coin, Key), u16, ValueQuery>;
+
+  #[pallet::genesis_build]
+  impl<T: Config> GenesisBuild<T> for GenesisConfig<T> {
+    fn build(&self) {
+      // Start at 1 to leave 0 available for a future design where Serai itself has a distinct
```

Comment:
This isn't considering "DPoS" a set. It's considering the Serai Tendermint process, if we break Serai Tendermint away from the Serai Bitcoin validators, a set. It's an idea we've discussed before. If we do make Serai itself have a distinct validator set, rather than using the sum of all others, I don't want it to be set no. 9 of all things, when it's trivial to reserve 0 for what likely is a scalability requirement.

Reply:
I'm not sure why Serai would need to be considered a set. We will always need one logical group that manages consensus around SRI, we can see this distinct and separate from the general sets that will form. Better to separate the concept in my opinion than try to reserve 0 in a combined concept.

Reply:
Right now, a super-set of all validators is used for Serai consensus. We've briefly discussed, due to limitations in Tendermint's scalability, solely using a dedicated set for Serai. We could use a pallet other than validator-sets? Except it'd still require the same bond mechanisms here. That's why I assumed we'd just re-use this pallet.

I'm not totally against using a dedicated pallet. That'd enable a few niceties...